lib/rucio/core/account.py

```diff
@@ -355,7 +356,7 @@ def del_account_attribute(account, key, *, session: "Session"):
 
 
 @read_session
-def get_usage(rse_id, account, *, session: "Session"):
+def get_usage(rse_id: uuid.UUID, account: InternalAccount, *, session: "Session") -> dict:
```

Comment:
I was looking at this statement within the function:
```
models.AccountUsage.rse_id == rse_id
```

Where the model is:

```
class AccountUsage(BASE, ModelBase):
    """Represents account usage"""
    __tablename__ = 'account_usage'
    account: Mapped[InternalAccount] = mapped_column(InternalAccountString(get_schema_value('ACCOUNT_LENGTH')))
    rse_id: Mapped[uuid.UUID] = mapped_column(GUID())
...
```

So I assumed `rse_id` would need to match the type indicated in the model

Reply:
Good assumption, but it’s complicated. You can see that SQLAlchemy will always return `Optional[str]` when querying the database:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/db/sqla/types.py#L70-L78

So in the API layer where the RSE name is converted to an RSE ID:

https://github.com/rucio/rucio/blob/cc0067cd2e5ae1c2c6c69cd8adc401db1bdcf015/lib/rucio/api/rse.py#L333-L334

It is most definitely a `str`.

Reply:
I changed all of them to `str`, it only caused this error:

```
Found 2 new problems in /home/runner/work/rucio/rucio/lib/rucio/core/account_limit.py
  - 1 errors with message """Expression of type "Unknown | UUID" cannot be assigned to declared type "str | None"""".
    Candidates: line 293
  - 1 errors with message """Argument of type "str | None" cannot be assigned to parameter "__key" of type "UUID" in function "get"
                               Type "str | None" cannot be assigned to type "UUID"
                                 "str" is incompatible with "UUID"""".
    Candidates: line 294
```

under `get_local_account_usage`. I've now annotated that `rse_id` to be `uuid.UUID`, and left all the others to be `str`.

Reply:
I have the impression that it would be more correct to change the model to have `rse_id: Mapped[str]`. The `GUID` class literally has `impl = CHAR`. I’m consulting the SQLAlchemy documentation, in hope of finding a definitive answer. Could you also investigate this path?

Reply:
> I have the impression that it would be more correct to change the model to have `rse_id: Mapped[str]`. The `GUID` class literally has `impl = CHAR`. I’m consulting the SQLAlchemy documentation, in hope of finding a definitive answer. Could you also investigate this path?

I'm not sure about making this change - looking at the other models, all `id` fields seem to use `uuid.UUID`. For consistency, I think it would be best to keep this one as `uuid.UUID` as well.

Reply:
If `Mapped[str]` is correct for `mapped_column(GUID())`, then all instances of `Mapped[uuid.UUID]` will have to be replaced.

Reply:
Addressed in https://github.com/rucio/rucio/pull/6497/commits/c20301e65f7f1a8c0c886beebdb4f17d01958d46

Reply:
Update - this causes `test_lifetime_creation_core` to fail:

```
E   psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "CONFIGS_PK"
E   DETAIL:  Key (section, opt)=(lifetime_model, cutoff_date) already exists.
```

Reply:
seems like it was a one-off issue - it passed when rerunning the test https://github.com/rucio/rucio/actions/runs/8005839275/job/21866803841?pr=6497#step:4:917